spec.html

```diff
@@ -49229,53 +49229,36 @@ <h1>
           </dl>
           <emu-alg>
             1. Let _values_ be a new empty List.
-            1. Let _remainingElementsCount_ be the Record { [[Value]]: 1 }.
+            1. Let _remainingElementsCount_ be 0.
             1. Let _index_ be 0.
             1. Repeat,
               1. Let _next_ be ? IteratorStepValue(_iteratorRecord_).
               1. If _next_ is ~done~, then
-                1. Set _remainingElementsCount_.[[Value]] to _remainingElementsCount_.[[Value]] - 1.
-                1. If _remainingElementsCount_.[[Value]] = 0, then
+                1. If _remainingElementsCount_ = 0, then
                   1. Let _valuesArray_ be CreateArrayFromList(_values_).
                   1. Perform ? Call(_resultCapability_.[[Resolve]], *undefined*, « _valuesArray_ »).
                 1. Return _resultCapability_.[[Promise]].
               1. Append *undefined* to _values_.
+              1. Set _remainingElementsCount_ to _remainingElementsCount_ + 1.
               1. Let _nextPromise_ be ? Call(_promiseResolve_, _constructor_, « _next_ »).
-              1. Let _steps_ be the algorithm steps defined in <emu-xref href="#sec-promise.all-resolve-element-functions" title></emu-xref>.
-              1. Let _length_ be the number of non-optional parameters of the function definition in <emu-xref href="#sec-promise.all-resolve-element-functions" title></emu-xref>.
-              1. Let _onFulfilled_ be CreateBuiltinFunction(_steps_, _length_, *""*, « [[AlreadyCalled]], [[Index]], [[Values]], [[Capability]], [[RemainingElements]] »).
+              1. Let _fulfilledSteps_ be a new Abstract Closure with parameters (_value_) that captures _values_, _resultCapability_, and _remainingElementsCount_ and performs the following steps when called:
+                1. Let _F_ be the active function object.
+                1. If _F_.[[AlreadyCalled]] is *true*, return *undefined*.
+                1. Set _F_.[[AlreadyCalled]] to *true*.
+                1. Let _index_ be _F_.[[Index]].
+                1. Set _values_[_index_] to _value_.
+                1. Set _remainingElementsCount_ to _remainingElementsCount_ - 1.
+                1. If _remainingElementsCount_ = 0, then
+                  1. Let _valuesArray_ be CreateArrayFromList(_values_).
+                  1. Return ? Call(_resultCapability_.[[Resolve]], *undefined*, « _valuesArray_ »).
+                1. Return *undefined*.
+              1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledSteps_, 1, *""*, « [[AlreadyCalled]], [[Index]] »).
               1. Set _onFulfilled_.[[AlreadyCalled]] to *false*.
               1. Set _onFulfilled_.[[Index]] to _index_.
```

Comment:
Why do these two need to remain internal slots on the function objects, rather than local variables like `_remainingElementsCount_` (but defined inside the loop)?

Reply:
Because these are not shared by all of the ACs defined in this AO. Remember this is happening in a loop.

Reply:
We should probably add a lint rule against an AC capturing an alias defined within a loop, just so we don't have to think about defining that.